docs/adr/0001-integration-of-event-store.md

```diff
@@ -0,0 +1,70 @@
+
+    The addition of a simple way to detect faulted instances when they occur would greatly improve the ability to proactively address problems in the workflow engine.
+
+These problems can be addressed by introducing the following features:
+- persistence of incoming instance creation requests and triggers before dispatching them to workflow instances.
+- persistence of "events" capturing the fact that a workflow transition has happened
+- creation of a "change feed" for the workflow engine, writing out generic events to cover workflow instance creation, transition, completion and faulting.
+
+It's worth noting that these features could be provided by rewriting Marain.Workflow as an event based system using the event sourcing approach. Whilst this would give us a foundation for addressing the above issues, it is a fundamental change that's best addressed by creating a new "version 2" of Marain.Workflow that's not backwards compatible with existing deployments.
+
+However one core part of an event sourcing-based system, namely the event store, provides a good mechanism for implementing the desired features:
+
+1. Conceptually, an event store is a good fit for the data we are recording. The data we wish to persist is essentially immutable* as it records things that have happened.
+2. A core part of an event store is a means to read events out of an "all stream" and process them. This lends itself to publishing these events as a change feed.
+
+\* "Essentially immutable" because there may be GDPR requirements that mean historical event data needs to be scrubbed of personally identifiable information at some point in the future.
+
+## Decision
+
+We will use Corvus.EventStore to persist events at three points in the system:
+
+1. When messages are received by the Message Processing service - "trigger received", "workflow instance creation request received".
+2. When a workflow creation request has been processed (or has failed to process correctly) - "workflow instance created", "workflow instance creation failed".
+3. When a trigger is processed by a workflow instance (or if the attempt to process the trigger failed and the workflow instance faulted) - "workflow instance transition completed", "workflow instance transition failed".
+
+The first of these forms our audit trail of incoming requests and will be written as events for a single "aggregate root", which is the Marain Tenant within which the requests take place.
+
+The second and third provide our history of change inside the workflow engine. For these, the workflow instance acts as our aggregate root.
+
+As always, storage will be siloed by tenant. This will require additional configuration to be supplied to the workflow engine per tenant to determine where event data will be stored.
+
+Further, we will publish these events out to Azure Event Grid as a change feed for the workflow engine. We will make this a configurable option.
```

Comment:
Is the intention that this will be a reliable stream of notifications? Are we intending to do the necessary work to handle transient failures correctly? Including the following:

1. Host process terminated (e.g., Azure decides to move us to different hardware) in between us deciding that a workflow state change has definitely occurred, and us reporting that fact to Event Grid
1. Event Grid temporarily unavailable for a few minutes
1. Longer term outages

The first would require us to implement any defining state change by making a persistent record, so that if we crash immediately afterwards, we can ensure that we go on to effect all of the necessary consequences during recovery. We'd probably also need to design for the assumption that we might send some notifications twice. (E.g., if our host process crashes in between us starting to send a message to Event Grid, and our being able to record a persistent record that Event Grid definitely received it, we have no way of knowing whether or not Event Grid actually received and forwarded our message. I don't _think_ Event Grid has intrinsic support for "exactly once" messages does it?)

The second would require us to decide whether a) we also go offline, or b) we design to be able to accommodate some level of backlog. (E.g,. as part of the persistence mechanism built to support the first one, do we manage a peristent queue of items that need to go out via Event Grid.)

The third might be accommodated by simply supporting an arbitrarily long backlog, but that might not be a good idea. If the intention is to provide the kind of dependability that we expect from, say, SQL Server, then we might simply need to go offline in cases like this.

Also, do we know what message ordering guarantees Event Grid offers? And if the answer is "none" do we need to do anything to ensure that consumers of messages can make sense of the world?

Reply:
The transient failure issue should be handled by the stream reader in Corvus.EventStore; the idea would be that we have a process separate to that which hosts the workflow engine, which reads events out of the all stream and publishes them as necessary. So if there was an outage between writing the event and publishing it out to event store, we would be able to recover that - the engine could keep processing triggers as normal, and the all stream reader would catch up and publish them out to event grid as soon as it was able.

If this potential delay between an event taking place and it being received via event grid were problematic for a subscriber, this could suggest that what they are doing might be better implemented as a workflow action.

As you suggest, our publishing mechanism based on the all stream cannot guarantee exactly-once or in-order delivery.